### ISSUE TITLE
Unbounded message re-queueing when TCP connection is dropped

### ISSUE THREAD

**Issue by igorsechyn** [nodemailer/nodemailer#1105]
**Node Version**: 12
**nodemailer Version**: 6.3.0
**OS**: ubuntu 16

How to reproduce:

- setup a `toxiproxy` in front of the mail server (we used `maildev` for testing)
- configure `toxiproxy` to drop connections after 15ms, e.g. `toxiproxy-cli toxic add mailsmtp -t timeout -a timeout=15` makes the tcp connection drop 15ms after it was opened.
- trigger sending emails through `nodemailer`

We observed following log messages:
```
[2020-02-14 21:09:05] INFO  [#2947] Created new pool resource #2947
[2020-02-14 21:09:05] DEBUG [#2947] Assigned message <d7b89eb0-bf92-9c00-917a-96b55e5c6bef@auth0user.net> to #2947 (1)
[2020-02-14 21:09:05] DEBUG [FLKP82ARUmA] Resolved toxiproxy as 10.12.13.9 [cache hit]
[2020-02-14 21:09:05] INFO  [FLKP82ARUmA] Connection established to 10.12.13.9:4007
[2020-02-14 21:09:05] DEBUG [FLKP82ARUmA] Closing connection to the server using "end"
[2020-02-14 21:09:05] INFO  [#2947] Connection #2947 was closed
[2020-02-14 21:09:05] INFO  [FLKP82ARUmA] Connection closed
[2020-02-14 21:09:05] DEBUG [#2947] Re-queued message <d7b89eb0-bf92-9c00-917a-96b55e5c6bef@auth0user.net> for #2947
[2020-02-14 21:09:05] INFO  [#2948] Created new pool resource #2948
[2020-02-14 21:09:05] DEBUG [#2948] Assigned message <d7b89eb0-bf92-9c00-917a-96b55e5c6bef@auth0user.net> to #2948 (1)
[2020-02-14 21:09:05] DEBUG [MPH0JMpPCc] Resolved toxiproxy as 10.12.13.9 [cache hit]
[2020-02-14 21:09:05] INFO  [MPH0JMpPCc] Connection established to 10.12.13.9:4007
[2020-02-14 21:09:05] DEBUG [MPH0JMpPCc] Closing connection to the server using "end"
[2020-02-14 21:09:05] INFO  [#2948] Connection #2948 was closed
[2020-02-14 21:09:05] INFO  [MPH0JMpPCc] Connection closed
[2020-02-14 21:09:05] DEBUG [#2948] Re-queued message <d7b89eb0-bf92-9c00-917a-96b55e5c6bef@auth0user.net> for #2948
[2020-02-14 21:09:05] INFO  [#2949] Created new pool resource #2949
```
Message was being requeued indefinitely until we turned off toxiproxy and connections stopped being dropped. While it is sensible to have this retry, I am worried that this may lead to high memory consumption and eventually causing a server using nodemailer to crash.

We observed this behaviour during a long outage of an SMTP provider, where connections were being dropped causing all messages being re-queued, so that internal nodemailer queue was growing very fast.

There are several options how this behaviour can be improved:
- adding a max number of retries
- adding an overall timeout for a message

But I would like to hear your opinions first, if you think this behaviour needs improvement and if so, what would be the best approach.

Thank you for the hard work on the library ♥️

**Comment by andris9** [comment]
Just to be sure - your actual problem is not requeing but the fact the message is kept in queue and during longer outage new messages are pushed to queue and as old ones are still there you might use up all memory?

For this exact situation there is “idle” event so you would not push new messages to the queue until old ones are cleared. https://nodemailer.com/smtp/pooled/#event-idle

**Comment by igorsechyn** [comment]
Hi @andris9,

thank you for your response.

> your actual problem is not requeing but the fact the message is kept in queue and during longer outage new messages are pushed to queue and as old ones are still there you might use up all memory?

Yes, this is my major concern

> For this exact situation there is “idle” event so you would not push new messages to the queue until old ones are cleared. https://nodemailer.com/smtp/pooled/#event-idle

I looked through the documentation, and IIUIC the suggestion is to pull from external queue, when there is at least one connection available. Unfortunately this approach has several drawbacks (especially in our use case):

* It basically limits the throughput of a worker handling messages from external queue to the number of connections in the pool
* Not all messages in external queue are processed by the same nodemailer transporter, e.g. we have transporters that talk to external services like SendGrid. In this case, there is no need to wait for an available connection in smtp pool to process this message
* A long outage would keep all available connections busy retrying the same message, which would cause external queue grow fast. This would cause violation of our SLOs for messages, that could have been processed by a different transporter

I think retrying the same message internally makes sense, however having an option to control how long or how often would provide more flexibility to client applications. In our case we would either reqeueu on external queue or discard it, making room for other messages to be processed.

**Comment by andris9** [comment]
Ok, I see your point. Could you make a PR with a proposed solution? It should probably limit the count of retries and return an error after all these retries fail.

**Comment by igorsechyn** [comment]
I will work on it. The change should not be too hard, however I am struggling to find a way of testing it, other than manually. I could not find any tests for existing behaviour either.

**Comment by igorsechyn** [comment]
I guess this issue can be closed now? Do you have a timeline for the next release?

**Comment by andris9** [comment]
Released as nodemailer@6.4.3
